### DISCUSSION TITLE
Explore Xarray and CF Xarray features related to CF Convention

### DISCUSSION THREAD

**Discussion by tomvothecoder** [xCDAT/xcdat#484] [3. Ideas]
@xCDAT/core-developers we should explore CF Xarray and Xarray features focused on CF Convention. We can think about citing useful features in the xCDAT documentation. **This exploration is important is because we want to make sure that we aren't duplicating functionalities that already exist and that first-time xCDAT/Xarray users are aware of them.** 

1. Xarray provides various CF features in the [`conventions.py` ](https://github.com/pydata/xarray/blob/main/xarray/conventions.py) file. 
    - [`decode_cf`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L489-L582)
    - [`cf_decoder`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L585-L624)
    - [`encode_dataset_coordinates`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L723-L742)
    - [`cf_encoder`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L745-L797)
    - [`decode_cf_variable`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L196-L292)
    - [`decode_cf_variables`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L374-L486)
    - [`encode_cf_variable`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L152-L193)
    - [`CF_RELATED_DATA`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L21-L32)
    - [`CF_RELATED_DATA_NEEDS_PARSING`](https://github.com/pydata/xarray/blob/d8ec3a3f6b02a8b941b484b3d254537af84b5fde/xarray/conventions.py#L33-L36)
 
2. CF Xarray, which is an xCDAT dependency, is a package focused on CF Convention. We use the `Dataset.cf.axes`, `xr.Dataset.cf.coordinates`, and `xr.Dataset.cf.bounds` attribute maps to interpret CF metadata in various parts of the xCDAT codebase.

    - Here is the link to the [API Reference](https://cf-xarray.readthedocs.io/en/latest/api.html) page.
    - One that stands out to me is the [`xarray.Dataset.cf.guess_coord_axis`](https://cf-xarray.readthedocs.io/en/latest/generated/xarray.Dataset.cf.guess_coord_axis.html#xarray.Dataset.cf.guess_coord_axis) API, which _"Automagically guesses X, Y, Z, T, latitude, longitude, and adds appropriate attributes. Uses regexes from Metpy and inspired by Iris function of same name."_ I definitely don't think xCDAT should call this behind the scenes, but I can see it being very useful for users to call in their own code in cases where datasets are not CF compliant.

**Comment by jypeter** [comment]
If somebody can investigate, I'll be **interested to know how `xarray` handles the longitude axis periodicity**.

When wondering about this and preparing my question, I have just discovered that I did not know either how CF handled the periodicity, and I have asked about it: [Can the longitude axis 360 degrees periodicity be described in CF ?](https://github.com/cf-convention/cf-conventions/issues/439)

Some of my earlier questions (in other issues) about periodicity assumed that periodicity metadata existed in CF. Theses questions may have seemed strange or obscure (if there is no way to describe periodicity). oops!
